File: cpp/tests/test_wav_to_csv.cpp
// cpp/tests/test_wav_to_csv.cpp
#include <gtest/gtest.h>
#include <fstream>
#include <vector>
#include <string>
#include <stdexcept>
#include <cstdlib>
#include <cstdint>
#include <sstream>

// Helper: write 16-bit PCM WAV
void writeWav16(const std::string& path, const std::vector<int16_t>& samples) {
    std::ofstream f(path, std::ios::binary);
    if (!f.is_open()) throw std::runtime_error("Failed to open file");

    uint32_t sampleRate = 16000;
    uint16_t channels = 1;
    uint16_t bitsPerSample = 16;
    uint32_t byteRate = sampleRate * channels * bitsPerSample / 8;
    uint16_t blockAlign = channels * bitsPerSample / 8;
    uint32_t dataSize = samples.size() * blockAlign;

    f.write("RIFF", 4);
    uint32_t chunkSize = 36 + dataSize;
    f.write(reinterpret_cast<const char*>(&chunkSize), 4);
    f.write("WAVE", 4);

    f.write("fmt ", 4);
    uint32_t subchunk1Size = 16;
    uint16_t audioFormat = 1; // PCM
    f.write(reinterpret_cast<const char*>(&subchunk1Size), 4);
    f.write(reinterpret_cast<const char*>(&audioFormat), 2);
    f.write(reinterpret_cast<const char*>(&channels), 2);
    f.write(reinterpret_cast<const char*>(&sampleRate), 4);
    f.write(reinterpret_cast<const char*>(&byteRate), 4);
    f.write(reinterpret_cast<const char*>(&blockAlign), 2);
    f.write(reinterpret_cast<const char*>(&bitsPerSample), 2);

    f.write("data", 4);
    f.write(reinterpret_cast<const char*>(&dataSize), 4);
    f.write(reinterpret_cast<const char*>(samples.data()), dataSize);
}

// Helper: write 24-bit PCM WAV
void writeWav24(const std::string& path, const std::vector<int32_t>& samples) {
    std::ofstream f(path, std::ios::binary);
    if (!f.is_open()) throw std::runtime_error("Failed to open file");

    uint32_t sampleRate = 16000;
    uint16_t channels = 1;
    uint16_t bitsPerSample = 24;
    uint32_t byteRate = sampleRate * channels * bitsPerSample / 8;
    uint16_t blockAlign = channels * bitsPerSample / 8;
    uint32_t dataSize = samples.size() * blockAlign;

    f.write("RIFF", 4);
    uint32_t chunkSize = 36 + dataSize;
    f.write(reinterpret_cast<const char*>(&chunkSize), 4);
    f.write("WAVE", 4);

    f.write("fmt ", 4);
    uint32_t subchunk1Size = 16;
    uint16_t audioFormat = 1; // PCM
    f.write(reinterpret_cast<const char*>(&subchunk1Size), 4);
    f.write(reinterpret_cast<const char*>(&audioFormat), 2);
    f.write(reinterpret_cast<const char*>(&channels), 2);
    f.write(reinterpret_cast<const char*>(&sampleRate), 4);
    f.write(reinterpret_cast<const char*>(&byteRate), 4);
    f.write(reinterpret_cast<const char*>(&blockAlign), 2);
    f.write(reinterpret_cast<const char*>(&bitsPerSample), 2);

    f.write("data", 4);
    f.write(reinterpret_cast<const char*>(&dataSize), 4);
    for (auto s : samples) {
        char bytes[3];
        bytes[0] = s & 0xFF;
        bytes[1] = (s >> 8) & 0xFF;
        bytes[2] = (s >> 16) & 0xFF;
        f.write(bytes, 3);
    }
}

// Helper: write float32 IEEE WAV
void writeWavFloat32(const std::string& path, const std::vector<float>& samples) {
    std::ofstream f(path, std::ios::binary);
    if (!f.is_open()) throw std::runtime_error("Failed to open file");

    uint32_t sampleRate = 16000;
    uint16_t channels = 1;
    uint16_t bitsPerSample = 32;
    uint32_t byteRate = sampleRate * channels * bitsPerSample / 8;
    uint16_t blockAlign = channels * bitsPerSample / 8;
    uint32_t dataSize = samples.size() * blockAlign;

    f.write("RIFF", 4);
    uint32_t chunkSize = 36 + dataSize;
    f.write(reinterpret_cast<const char*>(&chunkSize), 4);
    f.write("WAVE", 4);

    f.write("fmt ", 4);
    uint32_t subchunk1Size = 16;
    uint16_t audioFormat = 3; // IEEE float
    f.write(reinterpret_cast<const char*>(&subchunk1Size), 4);
    f.write(reinterpret_cast<const char*>(&audioFormat), 2);
    f.write(reinterpret_cast<const char*>(&channels), 2);
    f.write(reinterpret_cast<const char*>(&sampleRate), 4);
    f.write(reinterpret_cast<const char*>(&byteRate), 4);
    f.write(reinterpret_cast<const char*>(&blockAlign), 2);
    f.write(reinterpret_cast<const char*>(&bitsPerSample), 2);

    f.write("data", 4);
    f.write(reinterpret_cast<const char*>(&dataSize), 4);
    f.write(reinterpret_cast<const char*>(samples.data()), dataSize);
}

// Helper: load CSV
std::vector<float> loadCsv(const std::string& path) {
    std::ifstream f(path);
    if (!f.is_open()) throw std::runtime_error("Failed to open CSV file");

    std::string line;
    std::vector<float> samples;
    bool header = true;
    while (std::getline(f, line)) {
        if (header) { header = false; continue; }
        if (line.empty()) continue;
        auto pos = line.find(',');
        if (pos != std::string::npos) {
            float val = std::stof(line.substr(pos + 1));
            samples.push_back(val);
        }
    }
    return samples;
}

// Actual test
class WavToCsvTest : public ::testing::Test {
protected:
    std::vector<int16_t> pcm16 {0, 1000, -1000, 0, 32767, -32768, 0};
    std::vector<int32_t> pcm24 {0, 100000, -100000, 0, 500000, -500000, 0};
    std::vector<float>   float32 {0.0f, 0.5f, -0.5f, 0.0f, 1.0f, -1.0f, 0.0f};
};

TEST_F(WavToCsvTest, Pcm16ToCsv) {
    std::string wav = "test16.wav";
    std::string csv = "test16.csv";
    writeWav16(wav, pcm16);

    int ret = std::system(("./cpp/audio/wav_to_csv " + wav + " " + csv).c_str());
    ASSERT_EQ(ret, 0);

    auto got = loadCsv(csv);
    ASSERT_EQ(got.size(), pcm16.size());
    for (size_t i = 0; i < got.size(); i++) {
        if (pcm16[i] == 0)
            EXPECT_FLOAT_EQ(got[i], 0.0f);
        else
            EXPECT_NE(got[i], 0.0f);
    }
}

TEST_F(WavToCsvTest, Pcm24ToCsv) {
    std::string wav = "test24.wav";
    std::string csv = "test24.csv";
    writeWav24(wav, pcm24);

    int ret = std::system(("./cpp/audio/wav_to_csv " + wav + " " + csv).c_str());
    ASSERT_EQ(ret, 0);

    auto got = loadCsv(csv);
    ASSERT_EQ(got.size(), pcm24.size());
    for (size_t i = 0; i < got.size(); i++) {
        if (pcm24[i] == 0)
            EXPECT_FLOAT_EQ(got[i], 0.0f);
        else
            EXPECT_NE(got[i], 0.0f);
    }
}

TEST_F(WavToCsvTest, Float32ToCsv) {
    std::string wav = "test32.wav";
    std::string csv = "test32.csv";
    writeWavFloat32(wav, float32);

    int ret = std::system(("./cpp/audio/wav_to_csv " + wav + " " + csv).c_str());
    ASSERT_EQ(ret, 0);

    auto got = loadCsv(csv);
    ASSERT_EQ(got.size(), float32.size());
    for (size_t i = 0; i < got.size(); i++) {
        if (float32[i] == 0.0f)
            EXPECT_FLOAT_EQ(got[i], 0.0f);
        else
            EXPECT_NE(got[i], 0.0f);
    }
}

File: cpp/tests/test_wav_to_csv.py
# cpp/tests/test_wav_to_csv.py
import subprocess
import pandas as pd
import numpy as np
import wave
import struct
from pathlib import Path
import pytest


def write_float_wav(path: Path, samples: list[float]):
    """Write IEEE float32 WAV manually (since Python wave only supports PCM)."""
    sample_rate = 16000
    num_channels = 1
    bits_per_sample = 32
    audio_format = 3  # IEEE float
    byte_rate = sample_rate * num_channels * bits_per_sample // 8
    block_align = num_channels * bits_per_sample // 8
    subchunk2_size = len(samples) * block_align
    chunk_size = 36 + subchunk2_size

    with open(path, "wb") as f:
        # RIFF header
        f.write(b"RIFF")
        f.write(struct.pack("<I", chunk_size))
        f.write(b"WAVE")

        # fmt chunk
        f.write(b"fmt ")
        f.write(struct.pack("<I", 16))   # PCM fmt chunk size
        f.write(struct.pack("<H", audio_format))
        f.write(struct.pack("<H", num_channels))
        f.write(struct.pack("<I", sample_rate))
        f.write(struct.pack("<I", byte_rate))
        f.write(struct.pack("<H", block_align))
        f.write(struct.pack("<H", bits_per_sample))

        # data chunk
        f.write(b"data")
        f.write(struct.pack("<I", subchunk2_size))
        for s in samples:
            f.write(struct.pack("<f", s))


def write_test_wav(path: Path, bit_depth: int):
    """Generate test WAV with both zero and non-zero samples."""
    samples = [0, 1000, -1000, 0, 32767, -32768, 0]

    if bit_depth == 16:
        with wave.open(str(path), "wb") as wf:
            wf.setnchannels(1)
            wf.setsampwidth(2)
            wf.setframerate(16000)
            frames = struct.pack("<" + "h" * len(samples), *samples)
            wf.writeframes(frames)
        return np.array(samples, dtype=np.float32), 16

    elif bit_depth == 24:
        with wave.open(str(path), "wb") as wf:
            wf.setnchannels(1)
            wf.setsampwidth(3)
            wf.setframerate(16000)
            frames = b""
            for s in samples:
                frames += struct.pack("<i", s)[0:3]  # take low 3 bytes
            wf.writeframes(frames)
        return np.array(samples, dtype=np.float32), 24

    elif bit_depth == 32:
        floats = [float(s) / 32768.0 for s in samples]
        write_float_wav(path, floats)
        return np.array(floats, dtype=np.float32), 32

    else:
        raise ValueError("Unsupported bit depth")


def load_csv(csv_path: Path):
    """Load CSV samples from wav_to_csv output."""
    df = pd.read_csv(csv_path)
    assert "Sample" in df.columns
    return df["Sample"].to_numpy(dtype=np.float32)


@pytest.mark.parametrize("bit_depth", [16, 24, 32])
def test_wav_to_csv(tmp_path, bit_depth):
    wav_path = tmp_path / f"test_{bit_depth}.wav"
    csv_path = tmp_path / f"test_{bit_depth}.csv"

    expected, bd = write_test_wav(wav_path, bit_depth)

    result = subprocess.run(
        ["./cpp/audio/wav_to_csv", str(wav_path), str(csv_path)],
        capture_output=True,
        text=True,
    )
    assert result.returncode == 0, f"wav_to_csv failed: {result.stderr}"

    got = load_csv(csv_path)
    assert len(got) == len(expected)

    for i, (w, c) in enumerate(zip(expected, got)):
        if np.isclose(w, 0.0):
            assert np.isclose(c, 0.0), f"Mismatch at {i}: WAV={w}, CSV={c}"
        else:
            assert not np.isclose(c, 0.0), f"Unexpected zero at {i}: WAV={w}, CSV={c}"

